src/codemodder/codetf/v3/codetf.py

```diff
@@ -0,0 +1,181 @@
+    """Metadata describing fix outcome"""
+
+    status: FixStatus
+    reason: Optional[str]
+    details: Optional[str]
+
+
+class Rule(BaseModel):
+    id: str
+    name: str
+    url: Optional[str] = None
+
+
+class Finding(BaseModel):
+    id: str
+    rule: Optional[Rule] = None
+
+
+class Action(CaseInsensitiveEnum):
+    ADD = "add"
+    REMOVE = "remove"
+
+
+class PackageResult(CaseInsensitiveEnum):
+    COMPLETED = "completed"
+    FAILED = "failed"
+    SKIPPED = "skipped"
+
+
+class DiffSide(CaseInsensitiveEnum):
+    LEFT = "left"
+    RIGHT = "right"
+
+
+class PackageAction(BaseModel):
+    action: Action
+    result: PackageResult
+    package: str
+
+
+class Change(BaseModel):
+    lineNumber: int
+    description: Optional[str]
+    diffSide: DiffSide = DiffSide.RIGHT
+    properties: Optional[dict] = None
+    packageActions: Optional[list[PackageAction]] = None
+
+    @model_validator(mode="after")
+    def validate_lineNumber(self):
+        if self.lineNumber < 1:
+            raise ValueError("lineNumber must be greater than 0")
+        return self
+
+    @model_validator(mode="after")
+    def validate_description(self):
+        if self.description is not None and not self.description:
+            raise ValueError("description must not be empty")
+        return self
+
+
+class ChangeSet(BaseModel):
+    path: str
+    diff: str
+    changes: list[Change]
+
+
+class Reference(BaseModel):
+    url: str
+    description: Optional[str] = None
+
+    @model_validator(mode="after")
+    def validate_description(self):
+        self.description = self.description or self.url
+        return self
+
+
+class Strategy(Enum):
+    ai = "ai"
+    hybrid = "hybrid"
+    deterministic = "deterministic"
+
+
+class AIMetadata(BaseModel):
+    provider: Optional[str] = None
+    models: Optional[list[str]] = None
+    total_tokens: Optional[int] = None
+    completion_tokens: Optional[int] = None
+    prompt_tokens: Optional[int] = None
+
+
+class GenerationMetadata(BaseModel):
+    strategy: Strategy
+    ai: Optional[AIMetadata] = None
+    provisional: bool
+
+
+class FixMetadata(BaseModel):
+    # Fix provider ID, corresponds to legacy codemod ID
+    id: str
+    # A brief summary of the fix
+    summary: str
+    # A detailed description of the fix
+    description: str
+    references: list[Reference]
+    generation: GenerationMetadata
+
+
+class Rating(BaseModel):
+    score: int
+    description: Optional[str] = None
+
+
+class FixQuality(BaseModel):
+    safetyRating: Rating
+    effectivenessRating: Rating
+    cleanlinessRating: Rating
+
+
+class FixResult(BaseModel):
+    """Result corresponding to a single finding"""
+
+    finding: Finding
+    fixStatus: FixStatus
+    changeSets: list[ChangeSet]
+    fixMetadata: Optional[FixMetadata] = None
+    fixQuality: Optional[FixQuality] = None
+    # A description of the reasoning process that led to the fix
+    reasoningSteps: Optional[list[str]] = None
```

Comment:
Potentially each entry here should be something richer than a flat string: we might want to incorporate steps that represent user input/feedback.

Reply:
I could quibble that we have no idea what form that would take, and whether we should design for anything beyond a `str` now... but if you have short term thoughts that this is a promising data type, I'm okay with it.